### PR TITLE
COUCHDB-3224: Fixed database name encoding in view editor

### PR DESCRIPTION
**Pull request by sergey-safarov**

### CODE REVIEW

**Comment by garrensmith** [comment]
Thanks @sergey-safarov this is great. The tests are failing. Could you see why. And could you maybe add a test or two to prove this improves and fixes the `/` issues

**Comment by sergey-safarov** [comment]
Hello @garrensmith error is fixed

**Comment by sergey-safarov** [comment]
About test cases. I can make hotfix but test case creation is to complex task for me.

**Comment by garrensmith** [comment]
Ok. @robertkowalski needs to review this before we can merge.

**Comment by robertkowalski** [comment]
hey, thank you for your contribution -- i was on vacation, i'll check next week!

**Comment by robertkowalski** [comment]
@sergey-safarov the PR looks good, it will be superseeded by https://github.com/apache/couchdb-fauxton/pull/807 which contains more fixes and changes our integration tests to default to database names with special chars.

your commit will land seperately together with the other commits :)

**Comment by garrensmith** [comment]
@sergey-safarov thanks for trying to help us with this. We have decided to rather solve this with `encodeURIComponent`. I've just merged a PR that should fix the view section. 

Unfortunately we won't need this PR. But it would be great if you could test fauxton master and make sure we have now fixed all the encoding issues.

**Comment by sergey-safarov** [comment]
@garrensmith thank you for update
I tested current fauxton master all fine except one "Cancel" button.
To reproduce required start editing of view or create new view and press cancel.

**Comment by garrensmith** [comment]
Thanks for testing, would you be able to make a fix for the bug you found?

All misspelling thanks to my iPhone.




On Mon, Feb 27, 2017 at 5:31 PM +0200, "sergey-safarov" <notifications@github.com> wrote:












@garrensmith thank you for update

I tested current fauxton master all fine except one "Cancel" button.

To reproduce required start editing of view or create new view and press cancel.

**Comment by sergey-safarov** [comment]
Pull request is updated.

**Comment by garrensmith** [comment]
Could you rather do this work on a new PR and close this one. 

All misspelling thanks to my iPhone.




On Mon, Feb 27, 2017 at 6:13 PM +0200, "sergey-safarov" <notifications@github.com> wrote:












Pull request is updated.

**Comment by sergey-safarov** [comment]
@garrensmith I update pull request.

Please look at replication dialog. Generated POST with content
{"user_ctx":{"name":null,"roles":["_admin","_reader","_writer"]},"source":{"headers":{},"url":"http://217.12.247.220:5984/account%2F9b%2F44%2F9289fc701376b87ac2ad5ee78571"},"target":{"headers":{},"url":"http://217.12.247.220:5984/asdfsad"},"create_target":true,"continuous":false}

But replication is failed. This is CouchDB  issue?

File: app/addons/documents/doc-editor/components.react.jsx [hunk 1]
```diff
@@ -227,7 +227,7 @@ var AttachmentsPanelButton = React.createClass({
     var doc = this.props.doc.get('_id');
 
     return _.map(this.props.doc.get('_attachments'), function (item, filename) {
-      var url = FauxtonAPI.urls('document', 'attachment', db, doc, app.utils.safeURLName(filename));
+      var url = FauxtonAPI.urls('document', 'attachment', encodeURIComponent(db), doc, app.utils.safeURLName(filename));
```
Comment (garrensmith) [inline app/addons/documents/doc-editor/components.react.jsx]: Could you remove `safeURLName` we don't need it if we use the encodeURIComponent.

**Comment by sergey-safarov** [comment]
Requested changes in 04f05996264302a6802cbe6ca2ba4b8e33fcac93
If i correctly understand what is you requested.

Also i search for other places where is used "encodeURIComponent" and make changes in 51c209bd3e0031085599452db4c7f1f57832c77b

Please make review of this commits too.
If all is ok, then I will squash three into one commit.

**Comment by garrensmith** [comment]
Hi @sergey-safarov this is looking good. I spotted another encoding issue. If I create a new view and give it a `/` then it fails like this:


Would you be able to fix this as well? You should just need to do a decodeURIComponent in some of the react components. After that and once all the tests pass I can merge this in.

File: app/addons/documents/index-editor/stores.js [hunk 2]
```diff
@@ -186,7 +186,7 @@ Stores.IndexEditorStore = FauxtonAPI.Store.extend({
   getSaveDesignDoc: function () {
     if (this._designDocId === 'new-doc') {
       var doc = {
-        _id: '_design/' + this._newDesignDocName,
+        _id: '_design/' + encodeURIComponent(this._newDesignDocName),
```
Comment (sergey-safarov) [inline app/addons/documents/index-editor/stores.js]: Here is encoded DesignDoc name in POST request to CouchDB
Where is may be encoded viewname in  POST request to CouchDB?

File: app/addons/documents/components/jumptodoc.react.jsx [hunk 3]
```diff
@@ -28,7 +28,7 @@ const JumpToDoc = ({database, loadOptions}) => {
         ignoreCase={false}
         cache={false}
         onChange={({value: docId}) => {
-          const url = FauxtonAPI.urls('document', 'app', encodeURIComponent(database.id), app.utils.safeURLName(docId));
+          const url = FauxtonAPI.urls('document', 'app', encodeURIComponent(database.id), encodeURIComponent(docId));
```
Comment (sergey-safarov) [inline app/addons/documents/components/jumptodoc.react.jsx]: Is may be removed "safeURLName" function from app.utils?

File: app/addons/documents/index-editor/stores.js (hunk 2 above)
Comment (sergey-safarov) [inline app/addons/documents/index-editor/stores.js]: Is may be removed "safeURLName" function from app.utils?

**Comment by sergey-safarov** [comment]
Is may be removed "safeURLName" function from app.utils?

View name displaying is fixed view not work properly. I not know how to fix.

**Comment by sergey-safarov** [comment]
Hello @garrensmith 
Could you look error present in build log.
Think some dependency cannot be installed.

**Comment by garrensmith** [comment]
@sergey-safarov seems it was a travis issue. Could you fix the lint issues and try again.

File: app/addons/databases/actions.js [hunk 4]
```diff
@@ -155,7 +155,7 @@ export default {
           type: 'success',
           clear: true
         });
-        var route = FauxtonAPI.urls('allDocs', 'app', app.utils.safeURLName(databaseName), '?limit=' + Resources.DocLimit);
+        var route = FauxtonAPI.urls('allDocs', 'app', encodeURIComponent(databaseName), '?limit=' + Resources.DocLimit);
```
Comment (garrensmith) [inline app/addons/databases/actions.js]: Could you rather use this function https://github.com/apache/couchdb-fauxton/blob/master/app/addons/documents/base.js#L40 and then you don't need to encodeURIComponent

File: app/addons/databases/actions.js [hunk 5]
```diff
@@ -176,7 +176,7 @@ export default {
 
     databaseName = databaseName.trim();
 
-    const url = FauxtonAPI.urls('allDocs', 'app', app.utils.safeURLName(databaseName), '');
+    const url = FauxtonAPI.urls('allDocs', 'app', encodeURIComponent(databaseName), '');
```
Comment (garrensmith) [inline app/addons/databases/actions.js]: Same as aboce it seems its better to use `allDocsSantized` then you don't need to encodeURIComponent

**Comment by sergey-safarov** [comment]
@garrensmith I updated PR but exist two issues:
1) tests is looped
2) two tests is failed.

I can revert back last changes.

**Comment by garrensmith** [comment]
@sergey-safarov I will take a look next week. The PR has quite wide reaching effects. So I need to take a look at it. Leave it as is for now.

**Comment by wohali** [comment]
@garrensmith This looks like something that should get merged before 2.1.0 goes out. What else needs to be done to merge this? Did #807 or #813 fix this COUCHDB-3224 already?

**Comment by sergey-safarov** [comment]
@garrensmith i squashed all commits to one.
Can we all remaining changes to in other pull requests?

**Comment by garrensmith** [comment]
sorry @serdey-safarov, we are not going to merge this PR. We have fixed the encoding issues in other PR's. The encoding is a really difficult section and this PR has large changes and I'm not sure what effect they will have. I'm sorry that you PR was not accepted and I hope it won't stop you from contributing again

**Pull request closed without merging**